src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs

```diff
@@ -0,0 +1,421 @@
+            //explaining what (A) means
+            if (autoReferenceFound)
+            {
+                Console.WriteLine(Strings.ListPkg_AutoReferenceDescription);
+            }
+        }
+
+        /// <summary>
+        /// Fetches the latest versions for all of the packages that are
+        /// to be listed
+        /// </summary>
+        /// <param name="packages">The packages found in a project</param>
+        /// <param name="listPackageArgs">List args for the token and source provider</param>
+        /// <returns>A data structure like packages, but includes the latest versions</returns>
+        private async Task AddLatestVersionsAsync(
+            IEnumerable<FrameworkPackages> packages, ListPackageArgs listPackageArgs)
+        {
+            var providers = Repository.Provider.GetCoreV3();
+
+            foreach (var frameworkPackages in packages)
+            {
+                var topLevelTasks = new List<Task>();
+                foreach (var topLevelPackage in frameworkPackages.TopLevelPackages)
+                {
+                    topLevelTasks.Add(GetLatestVersionAsync(topLevelPackage, NuGetFramework.Parse(frameworkPackages.Framework), listPackageArgs, providers));
+                }
+
+                var transitiveTasks = new List<Task>();
+                foreach (var transitivePackage in frameworkPackages.TransitivePacakges)
+                {
+                    topLevelTasks.Add(GetLatestVersionAsync(transitivePackage, NuGetFramework.Parse(frameworkPackages.Framework), listPackageArgs, providers));
+                }
+
+                await Task.WhenAll(topLevelTasks);
+                await Task.WhenAll(transitiveTasks);
+            }
+        }
+
+        /// <summary>
+        /// Fetches the latest version of the given package
+        /// </summary>
+        /// <param name="package">The package to get the latest version for</param>
+        /// <param name="framework">The framework for the given package</param>
+        /// <param name="listPackageArgs">List args for the token and source provider></param>
+        /// <param name="providers">The providers to use when looking at sources</param>
+        /// <returns>An updated pacakge with the highest version at sources</returns>
+        private async Task GetLatestVersionAsync(
+            InstalledPackageReference package,
+            NuGetFramework framework,
+            ListPackageArgs listPackageArgs,
+            IEnumerable<Lazy<INuGetResourceProvider>> providers)
+        {
+            var sources = listPackageArgs.PackageSources;
+            var tasks = new List<Task>();
+
+            foreach (var packageSource in sources)
+            {
+                tasks.Add(GetLatestVersionPerSourceAsync(packageSource, listPackageArgs, package, framework, providers));
+            }
+
+            await Task.WhenAll(tasks);
+        }
+
+        /// <summary>
+        /// Gets the highest version of a package from a specific source
```

Comment:
What you are not adding here is the fact that this also updates the installed package reference package. 

You should not do that in this method. It's confusing. 

You can get the latest versions of all sources and then order once when they are all done.